File: amalthea/amalthea/src/wire/execute_request.rs
/*
 * execute_request.rs
 *
 * Copyright (C) 2022 by RStudio, PBC
 *
 */

use crate::wire::jupyter_message::MessageType;
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Represents a request from the front end to execute code
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ExecuteRequest {
    /// The code to be executed
    pub code: String,

    /// Whether the code should be executed silently (not shown to the user)
    pub silent: bool,

    /// Whether the code should be stored in history
    pub store_history: bool,

    /// Mapping of user expressions to be evaluated after code is executed.
    /// (TODO: should not be a plain value)
    pub user_expressions: Value,

    /// Whether to allow the kernel to send stdin requests
    pub allow_stdin: bool,

    /// Whether the kernel should discard the execution queue if evaluating the
    /// code results in an error
    pub stop_on_error: bool,
}

impl MessageType for ExecuteRequest {
    fn message_type() -> String {
        String::from("execute_request")
    }
}

File: extensions/kallichore-adapter/src/jupyter/ExecuteRequest.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (C) 2024 Posit Software, PBC. All rights reserved.
 *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
 *--------------------------------------------------------------------------------------------*/

import { JupyterChannel } from './JupyterChannel';
import { JupyterDisplayData } from './JupyterDisplayData';
import { JupyterRequest } from './JupyterRequest';


export class ExecuteRequest extends JupyterRequest<JupyterExecuteRequest, JupyterExecuteResult> {
	constructor(req: JupyterExecuteRequest) {
		super('execute_request', req, 'execute_result', JupyterChannel.Shell);
	}
}

/**
 * Represents an execute_request from the Jupyter frontend to the kernel.
 *
 * @link https://jupyter-client.readthedocs.io/en/stable/messaging.html#execute
 */
export interface JupyterExecuteRequest {
	/** The code to be executed */
	code: string;

	/** Whether the code should be executed silently */
	silent: boolean;

	/** Whether the code should be stored in history */
	store_history: boolean;                // eslint-disable-line

	/** A mapping of expressions to be evaluated after the code is executed (TODO: needs to be display_data) */
	user_expressions: Map<string, any>;    // eslint-disable-line

	/** Whether to allow the kernel to send stdin requests */
	allow_stdin: boolean;                  // eslint-disable-line

	/** Whether the kernel should stop the execution queue when an error occurs */
	stop_on_error: boolean;                // eslint-disable-line
}

/**
 * Represents an execute_result from the Jupyter kernel to the front end; this
 * is identical to the display_data message, with one additional field.
 *
 * @link https://jupyter-client.readthedocs.io/en/stable/messaging.html#id6
 */
export interface JupyterExecuteResult extends JupyterDisplayData {

	/** Execution counter, monotonically increasing */
	execution_count: number;  // eslint-disable-line
}
